### PR TITLE
Add Linkedin Konnector

### PR DESCRIPTION
**Pull request by Peltoche**
### Elements fetched:
- Fn / N
- Title
- Org
- Picture
- Linkedin Tag
- Linkedin Account
- Datapoint:
  - Phone Number
  - Email address
  - Linkedin Address
  - Address (all in locality input and the country specified)
  - Externe Website

All inputs should be RFC 6350 compliant.
All inputs are optionnal according the users inputs except:
- Fn / N
- Linkedin Tag
- Linedin Account

A merging system is also in place base on Google Contact tools (compare_contacts.coffee):
- Auto detect similar contacts
- Auto over-write the photo
- Complete the empty inputs and datapoints

### CODE REVIEW

File: server/konnectors/linkedin.coffee [hunk 1]
```diff
@@ -0,0 +1,328 @@
+
+# Vendor
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: Can you remove the vendor comments. It's a convention to write requirements first. So comments are useless here.

File: server/konnectors/linkedin.coffee [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+log = require('printit')
+    prefix: "Linkedin"
+    date: true
+
+ACCOUNT_TYPE = 'com.linkedin'
+
+# GLOBAL
+agent = 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:41.0) Gecko/20100101 Firefox/41.0'
+
+# Module
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: Can you put requirements before any assignation?

File: server/konnectors/linkedin.coffee [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+            .use(prepareCozyContacts)
+            .use(retrieveContacts)
+            .args(requiredFields, {}, {})
+            .fetch (err, fields, entries) ->
+                return callback err if err
+                log.info "Import finished"
+                callback()
+
+
+retrieveTokens = (requiredFields, entries, data, next) ->
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: Can you add to this function a comment explaining why and how you retrieve tokens?

File: server/konnectors/linkedin.coffee [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+        $ = cheerio.load body
+        entries.csrfToken = $('#loginCsrfParam-login').val()
+        entries.sourceAlias = $('#sourceAlias-login').val()
+        entries.accountName = requiredFields.login
+        log.info 'Tokens retrieved'
+        next()
+
+
+
+logIn = (requiredFields, entries, data, next) ->
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: Can you add to this function a comment explaining why and how you log in?

File: server/konnectors/linkedin.coffee [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+
+        if body == ""
+            log.info 'Login success'
+            next()
+        else
+            log.error 'Bad login / password'
+            next(new Error("Bad login/Password"))
+
+
+retrieveListContact = (requiredFields, entries, data, next) ->
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: Can you add to this function a comment explaining why and how you retrieve the contact list?

File: server/konnectors/linkedin.coffee [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+            next new Error(body.status_details)
+        entries.listContacts = body.contacts
+        if not entries.listContacts?
+            next new Error("Error retrieving contacts from request")
+        log.info 'List contact OK'
+        next()
+
+
+# Same as Google Konnector function
+prepareCozyContacts = (requiredFields, entries, data, next) ->
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: Can you add to this function a comment explaining why and how you retrieve Cozy contacts?

File: server/konnectors/linkedin.coffee [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+            if account?
+                entries.ofAccountByIds[account.id] = contact
+                entries.ofAccount.push contact
+        next()
+
+
+retrieveContacts = (requiredFields, entries, data, next) ->
+
+    processRetrievingContactData = (contact, next) ->
+        contacts = "https://www.linkedin.com/contacts/api/contacts/#{contact.id}/?fields=name,first_name,last_name,emails_extended,phone_numbers,sites,addresses,company,title,location,profiles,twitter,display_sources&_=1444864592009"
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: Lines with more than 80 chars are not allowed.

File: server/konnectors/linkedin.coffee [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+    opts =
+        url: 'https://linkedin.com'
+        jar: true
+        headers:
+            'User-Agent': agent
+
+    request.get opts, (err, res, body) ->
+        return next err if err
+
+        if body.status? and body.status == 'error'
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: With coffeescript, we use the keyworkd `is` instead of `==`.

File: server/konnectors/linkedin.coffee [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+            next new Error(body.status_details)
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: If an error occured, nothing should happened. So what follows should be displayed in a else block.

File: server/konnectors/linkedin.coffee [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+        url: contacts
+        jar: true
+        json: true
+        headers:
+            'User-Agent': agent
+
+    request.get opts, (err, res, body) ->
+        return next err if err
+        if body.status? and body.status == 'error'
+            next new Error(body.status_details)
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: Same when the `next` function is called, nothing should happen next.

File: server/konnectors/linkedin.coffee [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+                saveContact finalContact, entries
+
+
+
+
+
+getPhoneNumber = (data) ->
+    listPhones = []
+
+    data.phone_numbers.forEach (number) ->
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: You could use a `map` function here.

File: server/konnectors/linkedin.coffee [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+    endSavePicture = (err, updatedContact) ->
+        if err?
+            log.error "An error occured while creating or updating the " + \
+                      "contact."
+            log.raw err
+            done err
+
+
+        if not urlImage?
+            return
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: returns in the middle of a function are ugly.

File: server/konnectors/linkedin.coffee [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+
+getEmails = (data) ->
+    listEmails = []
+
+    data.emails_extended.forEach (email) ->
+        listEmails.push
+            name: 'email'
+            value: email.email
+            type: 'internet'
+            pref: email.primary == true ? true : undefined
```
Comment (nono) [inline server/konnectors/linkedin.coffee]: I'm surprised that coffeescript accepts this line. In fact, it does something really funny, but not what is wanted.

File: server/konnectors/linkedin.coffee [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+
+getAddresses = (data) ->
+    listAddresses = []
+
+    if data.location?
+        segmentAddress = data.location.split(', ')
+            .reverse()
+        country = segmentAddress[0] || ''
+        #region = segmentAddress[1] || ''
+        #locality = segmentAddress[2] || ''
```
Comment (nono) [inline server/konnectors/linkedin.coffee]: Why these lines are commented?

Comment (Peltoche) [inline server/konnectors/linkedin.coffee]: Currently the complete address is put in the locality field. The API authorize a way to define precisely some other fields but if we add them we will have some duplications with the complete address so we should wait an address parser before.

Comment (nono) [inline server/konnectors/linkedin.coffee]: @Peltoche OK, it's something that you should explain in a comment in your code ;-) The _why_ is often interesting and is difficult to infer from code.

File: server/konnectors/linkedin.coffee [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+        # Add all contact Id to the queue
+        entries.listContacts.forEach (contact) ->
+            queue.push contact.id, (err, finalContact) ->
+                log.error err if err
+                saveContact finalContact, entries
+
+
+
+
+
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: Put only 2 blank lines between functions.

File: server/konnectors/linkedin.coffee [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+    contacts = "https://www.linkedin.com/contacts/api/contacts/?start=0&count=10000&fields=id"
+    opts =
+        url: contacts
+        jar: true
+        json: true
+
+    request.get opts, (err, res, body) ->
+        return next err if err
+        if body.status? and body.status is 'error'
+            return next new Error(body.status_details)
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: Please do not use return in the middle of function, us `if` ... `else` block instead.

File: server/konnectors/linkedin.coffee [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+        cozyAccount = cozyContact.getAccount ACCOUNT_TYPE, entries.accountName
+        if ContactHelper.intrinsicRev(linkContact) isnt ContactHelper.intrinsicRev(cozyContact)
+            log.info "Update #{cozyContact.fn} from LinkedIn"
+            log.debug "Update #{linkContact.fn} from LinkedIn"
+            updateContact cozyContact, linkContact
+
+        else # Already uptodate, nothing to do.
+            log.info "LinkedIn contact #{cozyContact.fn} already synced and uptodate"
+            log.debug "LContact #{linkContact.n}already synced and uptodate"
+    else
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: Can you add a blank line above this `else`?

File: server/konnectors/linkedin.coffee [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+        for cozyContact in entries.cozyContacts
+            if CompareContacts.isSamePerson cozyContact, linkContact
+                log.info "Link #{cozyContact.fn} to linkedin account"
+                updateContact cozyContact, linkContact
+                return
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: This return is very ugly. Moreover, comparing each linkedin contacts with all Cozy contacts look very consuming. I think we should be more laxist about Contacts existence checking and only base it on name comparison. There is a duplicate feature in the Contact application that can handle duplicates later.

File: server/konnectors/linkedin.coffee [hunk 19]
```diff
@@ -0,0 +1,336 @@
...
+            log.info "Update #{cozyContact.fn} from LinkedIn"
+            log.debug "Update #{linkContact.fn} from LinkedIn"
+            updateContact cozyContact, linkContact
+
+        else # Already uptodate, nothing to do.
+            log.info "LinkedIn contact #{cozyContact.fn} already synced and uptodate"
+            log.debug "LContact #{linkContact.n}already synced and uptodate"
+
+    else
+        if entries.cozyContactsByFn[linkContact.fb]?
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: There is a typo here I think it's `fn` and not `fb`.

File: server/konnectors/linkedin.coffee [hunk 20]
```diff
@@ -0,0 +1,336 @@
...
+            log.info "Link #{cozyContact.fn} to linkedin account"
+            updateContact cozyContact, linkContact
+            return
```
Comment (frankrousseau) [inline server/konnectors/linkedin.coffee]: Here you should use `else` instead of `return`.

**Pull request merged**